src/nunit.analyzers/Extensions/INamespaceOrTypeSymbolExtensions.cs

```diff
@@ -0,0 +1,36 @@
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace NUnit.Analyzers.Extensions
+{
+    internal static class INamespaceOrTypeSymbolExtensions
+    {
+        /// <summary>
+        /// Checks the full name of a type or namespace without incurring the expense of building a full name string.
+        /// </summary>
+        public static bool HasFullName(this INamespaceOrTypeSymbol symbol, params string[] segments)
```

Comment:
Is possible to make it to accept string, like `"System.ComponentModel.TypeConverterAttribute"`? It would be easier to re-use it across solution, I think.

Reply:
I compared the Name properties rather than building a full name string because that's how Roslyn does it. If we start using Substring or Split, those allocations are just as bad and I'd rather just shoot for using the simplest code than implement that (maybe even Compilation.GetTypeByMetadataName).

Reply:
Sure, but I think it should possible to make it without Substring/Split and allocation-free (via Spans or direct char comparisons).

(I thought about using `Compilation.GetTypeByMetadataName` in such cases, but wasn't sure whether it is better, looked how it is implemented in roslyn-analyzers [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/Compiler/WellKnownTypeProvider.cs), and, well, it's not trivial there.)

Reply:
GetTypeByMetadataName is a little different. It doesn't recognize attributes that are declared internally inside assemblies. But this is really not a concern with TypeConverterAttribute since the BCL has provided this since the beginning.

Allocation-free is possible if the analyzer references System.Memory which runs slow span on .NET Framework (and VS will run the analyzer on .NET Framework). But it's also a lot of work. What's the motivation?

> (I thought about using Compilation.GetTypeByMetadataName in such cases, but wasn't sure whether it is better, looked how it is implemented in roslyn-analyzers here, and, well, it's not trivial there.)

This doesn't look like how it is used in Roslyn diagnostics. It's pretty much three ops: GetTypeByMetadataName, bail if null, otherwise use SymbolEqualityComparer.Default.Equals to compare the returned symbol to each attribute class.

Reply:
(If we start implementing alloc-free splitting, we should probably add tests. I'm not opposed but also I want to understand the gain versus replacing the whole thing with GetTypeByMetadataName.)

Reply:
> System.Memory which runs slow span on .NET Framework

Didn't know that, thanks! 
Still possible with direct chars comparison, but yeah, won't be pretty.

> What's the motivation?

We have lots of full name checks implemented via `GetFullMetadataName`. I'd like to be able to replace those with your method where possible.

Reply:
> If we start implementing alloc-free splitting

I was thinking more about "inline" comparison, not splitting, smth like
```csharp
            int index = fullName.Length - 1;
            if (!SegmentEquals(fullName, symbol.Name, ref index)) return false;
            
            ... (other segments checks)

            static bool SegmentEquals(string fullName, string segmentName, ref int index)
            {
                if (segmentName.Length > index + 1) return false;

                for (var i = segmentName.Length - 1; i >= 0; i--)
                    if (fullName[index--] != segmentName[i]) return false;

                return index < 0 || fullName[index--] == '.';
            }
```
But I agree that it's ugly.

>  the gain versus replacing the whole thing with GetTypeByMetadataName.

The only advantage I can think of is that we don't need `Compilation` for that to work, but that's probably not worth it. Less ugly code the better :)

Reply:
Should I go ahead and refactor this to use GetTypeByMetadata then?

Reply:
I think that would be great, thanks!

Reply:
Updated.